docs/language/informal/toplevel-inference.md

```diff
@@ -0,0 +1,360 @@
+treated as if it were annotated with dynamic.
+
+
+A field with no annotated type that does not override anything has the type
+inferred from its initializer.
+
+
+### Initializer inference
+
+
+Initializer inference works by defining a subset of the expressions in the
+language for which we can predictably do inference in a way that is not
+sensitive to ordering outside of the static dependencies that are syntactically
+apparent in the expression.  In particular, field inference cannot change the
+outcome of initializer inference.
+
+
+A meta-goal is to keep this inference relatively lightweight and efficient to
+implement.
+
+
+In all cases, if initializer inference fails, it is an error and produces no
+result.  Implementations are free to treat this as a result of dynamic after
+emitting the error for the purposes of emitting subsequent error messages, but
+the behavior of programs for which initializer inference fails is unspecified.
+
+#### Expressions with immediately-evident type
+
+We define a set of expressions with immediately-evident types (called
+immediately-evident expressions, abbreviated IE) as follows.  Immediately
+evident expressions have an inferred type as specified below.  Any expression
+which is not an IE has no inferred type.  For every immediately evident
+expression, we also define the set of variable dependencies upon which its
+inference relies.
+
+The intention is that toplevel inference should always return the same result as
+local inference, or else produce an error.
+
+
+* null, boolean, numeric, string, type, and symbol literals have their
+   corresponding type.
+   * No inference dependencies
+* An `await` of an expression has type `T` if the expression is an immediately
+evident expression with type `T` or `Future<T>`.
+   * Inference dependencies are the inference dependencies of the awaited
+     expression
+* A `throw` expression has type `bottom`.
+   * No inference dependencies
+* A parenthesized expression has the inferred type of the sub-expression
+   * Inference dependencies are the dependencies of the sub-expression
+* A conditional expression where both values are immediately evident expressions
+  has type `T` where `T` is the least upper bound of the inferred type of the
+  two returned sub-expressions.
+  * Inference dependencies are the union of the inference dependencies of the
+  two returned sub-expressions.
+* Logical boolean expressions and equality expressions have inferred type
+  `boolean`.
+  * No inference dependencies
+* Relational expressions, bitwise expressions, and shift expressions are treated
+as the appropriate method call as defined in the spec.
+  * Inference dependencies are those of the left hand operand.
+* A multiplicative binary expression is treated as a method call as defined in
+  the spec, with the same exceptions for when the left hand operand is of type `int`.
+  * Inference dependencies are those of the left hand operand.
```

Comment:
This doesn't work for `*` and `%`, since the exceptions for when the left hand operand is of type `int` require us to know the type of the right hand operand; we can't know this type reliably unless the right hand operand is also an interface dependency.

A similar situation exists for `+` and `-`, which it looks like you forgot to cover (I don't see a bullet addressing additive expressions).

Suggestion: change this to:

* A multiplicative or additive binary expression is treated as a method call as defined in the spec, with the same exceptions for when the left hand operand is of type `int`.
  * Inference dependencies are those of the left hand and right hand operands if the operator is `*`, `%`, `+`, or `-`.
  * Otherwise, inference dependencies are those of the left hand operand.